skill-tree/src/main/java/com/RDS/skilltree/Filters/JWTAuthenticationFilter.java

```diff
@@ -0,0 +1,64 @@
+package com.RDS.skilltree.Filters;
+
+import com.RDS.skilltree.Authentication.UserAuthenticationToken;
+import com.RDS.skilltree.utils.JWTUtils;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.util.WebUtils;
+
+import java.io.IOException;
+
+@Slf4j
+public class JWTAuthenticationFilter extends OncePerRequestFilter {
+
+    @Value("${cookieName}")
+    private String cookieName;
+    @Autowired
+    private JWTUtils jwtUtils;
+
+    @Override
+    public void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+
+        String token = getJWTFromRequest(request);
+
+
+            if (StringUtils.hasText(token) && jwtUtils.validateToken(token)) {
+                String rdsUserId = jwtUtils.getRDSUserId(token);
+                String role = jwtUtils.getUserRole(token);
```

Comment:
Rather than parsing the same `token` string multiple times, you can extract the Claims once, and then use the Claims to extract whatever information you need.

The `validateToken` method itself may be used to return the Claims if the token is valid.